test/writer.spec.js

```diff
@@ -63,21 +64,87 @@ describe('Writer', () => {
     })
   })
 
+  describe('applyProcessors', () => {
```

Comment:
I'll have to do the test review separately since I've recommended extracting this to an external file.